Added liveness and readiness checks to collector and query

jpkrohling opened the pull request:
The idea of this PR is to create a first version of the readiness and liveness checks. It can be improved in the future to expose the metrics from `expvar`, so that an orchestration service can auto scale based on the data reported by this.

Similarly, this would allow the server to keep booted even if the underlying data store is offline: while the main HTTP port might still accept incoming requests and put on a buffer, the health check endpoint might return a `degraded` state, so that admins could get notified if something is wrong.

Sample output for "ready and alive":
```
curl -v http://localhost:17686/
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 17686 (#0)
> GET / HTTP/1.1
> Host: localhost:17686
> User-Agent: curl/7.53.1
> Accept: */*
> 
< HTTP/1.1 204 No Content
< Date: Thu, 20 Jul 2017 12:18:20 GMT
< 
* Connection #0 to host localhost left intact
```

Sample output for "booted, but in trouble" (ie: not passing liveness check):
```
curl -v http://localhost:15268/
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 15268 (#0)
> GET / HTTP/1.1
> Host: localhost:15268
> User-Agent: curl/7.53.1
> Accept: */*
> 
< HTTP/1.1 500 Internal Server Error
< Date: Thu, 20 Jul 2017 12:19:38 GMT
< Content-Length: 20
< Content-Type: text/plain; charset=utf-8
< 
* Connection #0 to host localhost left intact
Server not available
```

Sample output for "booting" (ie: not passing readiness check):
```
curl -v http://localhost:15268/
*   Trying ::1...
* TCP_NODELAY set
* Connected to localhost (::1) port 15268 (#0)
> GET / HTTP/1.1
> Host: localhost:15268
> User-Agent: curl/7.53.1
> Accept: */*
> 
< HTTP/1.1 503 Service Unavailable
< Date: Thu, 20 Jul 2017 12:20:12 GMT
< Content-Length: 20
< Content-Type: text/plain; charset=utf-8
< 
* Connection #0 to host localhost left intact
Server not available
```

jpkrohling commented on cmd/collector/app/builder/builder_flags.go:
```diff
@@ -38,4 +38,6 @@ var (
 	// CollectorHTTPPort is the port that the collector service listens in on for http requests
 	CollectorHTTPPort = flag.Int("collector.http-port", 14268, "The http port for the collector service")
+	// HealthCheckHTTPPort is the port that the health check service listens in on for http requests
+	HealthCheckHTTPPort = flag.Int("health-check.http-port", 15268, "The http port for the health check service")
```
Like most of the code on this PR, this is duplicated, with one version for the collector and one for the query. This is because we might want to evolve the health checks for each module to be unique, as the query might have different concerns as the collector: the collector might still ingest data and put it on a temp buffer, while the query might not even be able to retrieve data to the UI.

jpkrohling replied:
In environments like OpenShift and Kubernetes, it's easier to protect ports than paths. For instance, this port could be kept for internal communication only, while the main HTTP port can be exposed as a "route".

jpkrohling commented on cmd/collector/main.go:
```diff
@@ -48,10 +51,30 @@ const (
+		// Start the health check server, used for reporting the liveness and readiness state for this component
+		logger.Info("Health Check server started", zap.Int("http-port", *builder.HealthCheckHTTPPort))
+		healthCheckMux := http.NewServeMux()
+		healthCheckMux.HandleFunc("/", func(w http.ResponseWriter, _ *http.Request) {
```
For this initial PR, this might be OK, but it would probably be better in a next iteration to extract this into its own source file.

yurishkuro replied:
+1, this is too much code for `main()`

per my comment above about not defining a separate port, and considering that http server is starting after the tchannel server, I think all you really need is a very simple `/health` handler that always returns `OK`.

jpkrohling replied:
The advantage of having its own mux is that it won't influence the main mux in any way (metrics, concurrency, ...). Also, if the main mux is stuck/hung for some reason, the healthcheck mux is still able to report the server's status.

jpkrohling commented on cmd/collector/main.go:
```diff
@@ -85,10 +108,27 @@ func main() {
+
+	select {
+	case <-serverChannel:
+		logger.Info("Jaeger Collector is finishing")
```
HEADS UP: this change means that stopping the server now ends up with a return code 0. Previously, the return code was 130:

Before this PR:
```
$ ./cmd/collector/collector-linux
{"level":"info","ts":1500553817.5565636,"caller":"collector/main.go:90","msg":"Listening for HTTP traffic","http-port":14268}
^C
$ echo $?
130
```

After this PR:
```
$ ./cmd/collector/collector-linux
{"level":"info","ts":1500553862.3376029,"caller":"collector/main.go:62","msg":"Health Check server started","http-port":15268}
{"level":"info","ts":1500553862.3593538,"caller":"collector/main.go:114","msg":"Starting Jaeger Collector HTTP server","http-port":14268}
^C{"level":"info","ts":1500553864.2055447,"caller":"collector/main.go:132","msg":"Jaeger Collector is finishing"}
$ echo $?
0
```

yurishkuro replied:
also, what are we gaining by trapping the signals vs. how it was before with a permanent block on `select{}`?

jpkrohling replied:
Yes: OpenShift / Kubernetes might alert admins if a container exists with a non-zero return code. 

Trapping the kill signal allows for an orderly shutdown of the server, so that inflight requests can be terminated and connections to downstream services can be properly closed. This might not be a big issue on an environment where there are only a few instances of the services, but on a highly elastic environment, pods are created and destroyed quite frequently.

yurishkuro replied:
the reason you may have received panic was because you were not synchronizing the shared variable. I think channel-based implementation requires more code than an atomic.Int var that you Store and Load.

What I don't like the most in the current version is that it's too verbose, like 20-30 lines of code added to main (which is also untested code since main aren't tested), instead of just a couple of lines to start health server and to update it to Ready state.

jpkrohling replied:
It's more like 15 lines, but I see your point. I think this is because we have two different things going on here: the first is to make the current HTTP server to start in a go func, blocking the main thread via a channel, to allow for an orderly shut down. The second is the actual health check server, which is started in a similar way.

Would it make sense to try to refactor this code, so keep it clean and perhaps even adding some tests?

yurishkuro replied:
yes, refactor to minimize the amount of code it adds to main, since we will want to reuse it in the other binaries. Perhaps it can even be written in a functional style, like

```go
health.Start(*port, func() {
    // the rest of main
})
```

Have you checked how Kubernetes deals with that?

jpkrohling replied:
I checked both OpenShift and Kubernetes and they work on top of Cobra. So, a Cobra command for "all in one" (OpenShift here), would call the "StartMaster" command, which would then start that component.

In our case, if we were to take OpenShift as example, it would look like this:

- Single binary for the whole application (not required, but I see some advantages to it)
- Commands passed to the binary specify what to do (`jaeger start-all-in-one` or even `jaeger start all-in-one`)
- The Cobra's "`Run`" for the top-level commands would invoke specialized commands, like, `jaeger start all-in-one` means `pkg/cmd/hcOptions.RunHealthCheckServer` + `pkg/cmd/collectorOptions.RunCollector` + `pkg/cmd/agentOptions.RunAgent`, ... (reference: https://github.com/openshift/origin/blob/master/pkg/cmd/server/start/start_allinone.go#L286) .
- With this, the main file itself is quite clean: https://github.com/openshift/origin/blob/master/cmd/openshift/openshift.go

jpkrohling replied:
Considering that @pavolloffay is doing some Cobra work in another PR, I'd see if we can merge this PR as it is, and refactor the whole bootstrap procedure in a follow up PR. We can then discuss how to architect it, perhaps using OpenShift / Kubernetes as a model.

What do you think?

jpkrohling commented:
Looks like I need to do some better testing before sending PRs :) 

```
/tmp/go-build859975043/github.com/uber/jaeger/cmd/standalone/_test/standalone.test flag redefined: health-check.http-port
```

yurishkuro commented on cmd/query/main.go:
```diff
@@ -68,13 +91,31 @@ func main() {
+
+	select {
+	case <-serverChannel:
+		logger.Info("Jaeger Query is finishing")
```
if we end up keeping the signal trapping approach (I'm not seeing the value just yet), then it makes more sense for the log line to say something like "jaeger-query interrupted & exiting"

jpkrohling commented:
@yurishkuro , is this feature something you'd consider merging? If so, which changes would be needed, other than having the health check server on its own source files?

yurishkuro commented:
Yes, it makes sense to merge. A few thoughts:
  * I would make it a separate shared component, somewhere in /pkg/health, and reuse in collector/agent/query
  * it should be a struct that can be created in main() and have methods to start serving and to change status
  * at Uber we usually handle SIGTERM by changing the health state to "unavailable" first, and then existing after a short delay. It allows health endpoint to be used in traffic controllers so that it will stop sending requests to the service, thus allowing it to drain/finish in flight requests before exiting

jpkrohling commented:
I just updated this PR to implement the first two points (except that I didn't do it for the agent).

> at Uber we usually handle SIGTERM by changing the health state to "unavailable" first, and then existing after a short delay. It allows health endpoint to be used in traffic controllers so that it will stop sending requests to the service, thus allowing it to drain/finish in flight requests before exiting

I understand that you have your own `main.go` file, so, is this something we need to handle here? If so, I'm not quite sure I understand what has to be done: should I change the status when receiving a signal and let the code "sleep" for some time? From what I understand, that wouldn't help that much, as the HTTP server would already be down, no?

jpkrohling commented on cmd/collector/app/builder/builder_flags.go:
```diff
@@ -38,4 +38,6 @@ var (
 	// CollectorHTTPPort is the port that the collector service listens in on for http requests
 	CollectorHTTPPort = flag.Int("collector.http-port", 14268, "The http port for the collector service")
+	// CollectorHealthCheckHTTPPort is the port that the health check service listens in on for http requests
+	CollectorHealthCheckHTTPPort = flag.Int("collector.health-check-http-port", 15268, "The http port for the health check service")
```
No reason, I just added 1000 to the other HTTP port. I could add +1 instead of +1000, though.

jpkrohling commented on cmd/query/main.go:
```diff
@@ -35,15 +38,26 @@ import (
+	h := healthcheck.Handler{State: &state, Port: *builder.QueryHealthCheckHTTPPort, Logger: *logger}
+	go func() {
+		h.Register()
+	}()
```
Not right now. If we do need it in the future, we can refactor it accordingly.

jpkrohling commented on cmd/query/main.go:
```diff
@@ -68,13 +82,31 @@ func main() {
+		// between allocating a port and serving a request.
+		state = http.StatusNoContent
+
+		logger.Fatal("Could not launch service", zap.Error(http.ListenAndServe(httpPortStr, recoveryHandler(r))))
```
OK, I've seen some other code around (gopl.io, I think) where they used this kind of notation, but I also think your version is more readable .

jpkrohling commented on cmd/query/main.go:
```diff
@@ -68,13 +82,31 @@ func main() {
+		// there's still a chance that a request comes in between the `ready` flag and the server's
+		// actual startup, but this would be a rare situation, as we assume the `ListenAndServe` is fast enough
+		// between allocating a port and serving a request.
+		state = http.StatusNoContent
```
I ended up implementing as channels, as we are talking about a go channel calling a method on an object set by another go channel. When developing, I got a panic in one situation, as the handler was still null when Update was called. With channels, it works as expected.

Query main HTTP server started before:
```json
{"level":"info","ts":1501236900.1643984,"caller":"query/main.go:56","msg":"Ticking"}
{"level":"info","ts":1501236900.187948,"caller":"query/main.go:99","msg":"Starting Jaeger Query HTTP server","port":16686}
{"level":"info","ts":1501236901.164993,"caller":"healthcheck/handler.go:48","msg":"Health Check server started","http-port":16687}
{"level":"info","ts":1501236901.1652951,"caller":"healthcheck/handler.go:68","msg":"Health Check state change","new-state":204}
```

Health check HTTP server started before:
```json
{"level":"info","ts":1501236874.302359,"caller":"healthcheck/handler.go:49","msg":"Health Check server started","http-port":16687}
{"level":"info","ts":1501236874.3137684,"caller":"query/main.go:94","msg":"Starting Jaeger Query HTTP server","port":16686}
{"level":"info","ts":1501236875.302688,"caller":"healthcheck/handler.go:72","msg":"Health Check state change","new-state":204}
```

PS: I tested it manually with a ticker:

```go
tick := time.Tick(1 * time.Second)
logger.Info("Ticking")
<-tick
```

jrbury commented on cmd/collector/main.go:
```diff
@@ -85,10 +101,27 @@ func main() {
+
+	select {
+	case <-serverChannel:
+		logger.Info("Jaeger Collector is finishing")
```
I'm not sure if you're still working on this part (let me know if you are and you can ignore me) but with the current way this select is done it will receive the signal and main will instantly exit, preventing the safe shutdown. Adding a waitgroup is the most common solution in this approach.

jpkrohling replied:
Right, that's true. The idea is to prepare the current code for next PRs, which would then add the actual orderly shutdown logic.

jpkrohling commented:
PR updated to fix the merge conflicts. Once this is merged, I'll send a plan to refactor the bootstrap code, to be more testable and reusable.

yurishkuro commented on cmd/query/main.go:
```diff
@@ -57,6 +64,13 @@ func main() {
+					logger.Fatal("Could not start the health check server.", zap.Error(err))
+				}
+			}()
+
```
I still have the issue with the verbosity of it. Most of the logic can be moved inside the healthChecker itself, e.g. main() does not need to know anything about the hcChannel. Literally, the usage can be reduced to this:

```go
healthServer, err := healthcheck.Serve(queryOpts.QueryHealthCheckHTTPPort, logger)
if err != nil {
    logger.Fatal("Could not start the health check server.", zap.Error(err))
}
```

and once everything is started

```go
healthServer.Ready()
```

the http status codes can also be hidden inside the implementation.

yurishkuro commented on cmd/query/main.go:
```diff
@@ -89,9 +103,27 @@ func main() {
+				hcChannel <- http.StatusNoContent
+
+				logger.Info("Starting jaeger-query HTTP server", zap.Int("port", queryOpts.QueryPort))
+				if err := http.ListenAndServe(portStr, recoveryHandler(r)); err != nil {
```
why not start the main server _before_ setting the healthServer to ready? There are other ports that each service might be listening to, I don't think any one of them needs to be made special.

jpkrohling replied:
+1 . Should this be outside of the go routine, before the select on the channel? There's a chance that the health check server would be marked as ready before the main server starts, but it should not be such a big issue, I think.

yurishkuro replied:
there is no bullet-proof way to avoid that aside from having the healthcheck to actually ping the other ports opened by the service. Which could be a future enhancement of it, e.g. by passing "ping actors" to the Ready() functions, e.g.

```go
hc.Ready(
    healthcheck.HTTPChecker(apiPort),
    healthcheck.HTTPChecker(zipkinPort),
    healthcheck.TChannelChecker(collectorPort),
)
```

yurishkuro commented on cmd/collector/main.go:
```diff
@@ -42,12 +45,16 @@ import (
 )
 
 func main() {
+	var serverChannel = make(chan os.Signal, 0)
```
nit: signalsChannel

yurishkuro commented on pkg/healthcheck/handler.go:
```diff
@@ -0,0 +1,82 @@
+	hcPortStr := ":" + strconv.Itoa(h.Port)
+
+	go func() {
+		if err := http.ListenAndServe(hcPortStr, healthCheckMux); err != nil {
```
this file should have tests. Once you start doing tests you will find out that you want a Stop or Close method on the Handler. You can see how that's done in the agent's HTTP server.

yurishkuro commented on cmd/query/app/builder/builder_flags.go:
```diff
@@ -40,19 +41,24 @@ type QueryOptions struct {
 	flagSet.String(queryPrefix, "api", "The prefix for the url of the query service")
 	flagSet.String(queryStaticFiles, "jaeger-ui-build/build/", "The path for the static assets for the UI")
+	flagSet.Int(queryHealthCheckHTTPPort, 16687, "The http port for the health check service")
+
```
rm line

jpkrohling commented:
I just updated the PR with some tests. Although the tests would pass fine without the `Close()` method on the handler, I added it anyway, as I don't like the idea of just letting processes to linger on random ports, even if just during the tests :)

yurishkuro commented:
We need to investigate why 1.8 is so shower. Meanwhile I'd go with splitting the server and listener in the health server (it does not need to be exposed to main).

jpkrohling commented:
> Coverage decreased (-0.03%) to 99.968% 

It's because of one line that is reached only when `http.Serve` returns. I couldn't find a reasonable way to test it on 1.7.

yurishkuro commented:
I think if you stop/close the listener the Serve method exists with the error. You have similar code in the Agent, and its test has 100% coverage.

yurishkuro commented on pkg/healthcheck/handler.go:
```diff
@@ -0,0 +1,100 @@
+}
+
+// NewState creates a new state instance. The initial state is what's specified with the state parameter
+func NewState(state int, logger *zap.Logger) (*State, error) {
```
Should not be returning error since it never does
